File: packages/server/backbone/kotlin/src/main/kotlin/io/github/hieudoanm/backbone/cache/Cache.kt
package io.github.hieudoanm.backbone.cache

import io.github.hieudoanm.backbone.core.AppConfig
import io.github.hieudoanm.backbone.database.Database
import io.github.hieudoanm.backbone.database.toList
import kotlinx.coroutines.*
import java.util.concurrent.ConcurrentHashMap

class CacheStore(private val config: AppConfig, private val db: Database) {
    private data class Entry(
        val value: String,
        val ttl: Long,
        val expiresAt: Long,
    )

    private val store = ConcurrentHashMap<String, Entry>()
    private var scope: CoroutineScope? = null

    fun startEviction(scope: CoroutineScope) {
        this.scope = scope
        scope.launch {
            while (isActive) {
                delay(30_000)
                val now = System.currentTimeMillis()
                store.entries.removeIf { (_, e) -> e.ttl > 0 && e.expiresAt <= now }
            }
        }
        loadFromDb()
    }

    private fun loadFromDb() {
        db.connection().use { conn ->
            val rows = conn.createStatement().executeQuery("SELECT * FROM _cache").toList()
            for (row in rows) {
                val key = row["key"] as String
                val value = row["value"] as String
                val ttl = (row["ttl"] as? Number)?.toLong() ?: 0L
                val expiresAtStr = row["expires_at"] as? String ?: ""
                val expiresAt = if (expiresAtStr.isNotBlank()) {
                    try {
                        java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss").parse(expiresAtStr).time
                    } catch (_: Exception) { 0L }
                } else 0L
                if (ttl > 0 && expiresAt <= System.currentTimeMillis()) continue
                store[key] = Entry(value, ttl, expiresAt)
            }
        }
    }

    fun get(key: String): Map<String, Any?>? {
        val entry = store[key] ?: return null
        if (entry.ttl > 0 && entry.expiresAt <= System.currentTimeMillis()) {
            store.remove(key)
            return null
        }
        return mapOf("key" to key, "value" to entry.value, "ttl" to entry.ttl)
    }

    fun set(key: String, value: String, ttlSeconds: Long = 0) {
        val expiresAt = if (ttlSeconds > 0) System.currentTimeMillis() + ttlSeconds * 1000 else 0L
        store[key] = Entry(value, ttlSeconds, expiresAt)
        val expiresAtStr = if (ttlSeconds > 0) {
            java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(java.util.Date(expiresAt))
        } else ""
        db.connection().use { conn ->
            conn.createStatement().executeUpdate("""
                INSERT OR REPLACE INTO _cache (key, value, ttl, expires_at, created_at, updated_at)
                VALUES ('${key.replace("'", "''")}', '${value.replace("'", "''")}',
                        $ttlSeconds, '$expiresAtStr', datetime('now'), datetime('now'))
            """)
        }
    }

    fun delete(key: String): Boolean {
        val removed = store.remove(key)
        db.connection().use { conn ->
            conn.createStatement().executeUpdate(
                "DELETE FROM _cache WHERE key = '${key.replace("'", "''")}'"
            )
        }
        return removed != null
    }

    fun flush() {
        store.clear()
        db.connection().use { conn ->
            conn.createStatement().executeUpdate("DELETE FROM _cache")
        }
    }

    fun list(): List<Map<String, Any?>> {
        val now = System.currentTimeMillis()
        return store.entries.map { (key, entry) ->
            val valid = entry.ttl <= 0 || entry.expiresAt > now
            val expiresAtStr = if (entry.expiresAt > 0) {
                java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss").format(java.util.Date(entry.expiresAt))
            } else ""
            mapOf(
                "key" to key,
                "value" to (if (valid) entry.value else "(expired)"),
                "ttl" to entry.ttl,
                "expires_at" to expiresAtStr,
            )
        }
    }

    fun stats(): Map<String, Any> {
        val now = System.currentTimeMillis()
        val total = store.size
        val expired = store.values.count { it.ttl > 0 && it.expiresAt <= now }
        return mapOf("total_entries" to total, "expired_entries" to expired)
    }
}

File: packages/app/headless/backbone/languages/go/internal/cache/cache.go
// Package cache provides an in-memory cache that persists to SQLite.
package cache

import (
	"database/sql"
	"log"
	"sync"
	"time"

	"github.com/hieudoanm/backbone/internal/store"
)

// CacheStore is a thread-safe key/value cache backed by the database.
type CacheStore struct {
	mu   sync.RWMutex
	data map[string]*store.CacheEntry
	db   *sql.DB
}

// NewCacheStore creates a cache, warm it from disk, and starts eviction.
func NewCacheStore(db *sql.DB) *CacheStore {
	cs := &CacheStore{
		data: make(map[string]*store.CacheEntry),
		db:   db,
	}
	cs.loadFromDB()
	go cs.evictLoop()
	return cs
}

func (cs *CacheStore) loadFromDB() {
	rows, err := cs.db.Query(`SELECT key, value, ttl, expires_at, created_at, updated_at FROM _cache`)
	if err != nil {
		log.Printf("cache: load from db: %v", err)
		return
	}
	defer rows.Close()
	for rows.Next() {
		var e store.CacheEntry
		if err := rows.Scan(&e.Key, &e.Value, &e.TTL, &e.ExpiresAt, &e.CreatedAt, &e.UpdatedAt); err != nil {
			log.Printf("cache: scan row: %v", err)
			continue
		}
		if e.TTL > 0 && e.ExpiresAt != "" {
			expires, err := time.Parse(time.RFC3339, e.ExpiresAt)
			if err == nil && time.Now().UTC().After(expires) {
				continue
			}
		}
		cs.data[e.Key] = &e
	}
}

func (cs *CacheStore) evictLoop() {
	ticker := time.NewTicker(30 * time.Second)
	for range ticker.C {
		now := time.Now().UTC()
		cs.mu.Lock()
		for key, entry := range cs.data {
			if entry.TTL > 0 && entry.ExpiresAt != "" {
				expires, err := time.Parse(time.RFC3339, entry.ExpiresAt)
				if err == nil && now.After(expires) {
					delete(cs.data, key)
				}
			}
		}
		cs.mu.Unlock()
	}
}

// Set stores a value, optionally expiring after ttl seconds (0 = never).
func (cs *CacheStore) Set(key, value string, ttl int) *store.CacheEntry {
	now := time.Now().UTC()
	nowStr := now.Format(time.RFC3339)
	var expiresAt string
	if ttl > 0 {
		expiresAt = now.Add(time.Duration(ttl) * time.Second).Format(time.RFC3339)
	}
	entry := &store.CacheEntry{
		Key:       key,
		Value:     value,
		TTL:       ttl,
		ExpiresAt: expiresAt,
		CreatedAt: nowStr,
		UpdatedAt: nowStr,
	}
	cs.mu.Lock()
	cs.data[key] = entry
	cs.mu.Unlock()
	cs.persist(entry)
	return entry
}

// Get fetches a value, honoring TTL expiry.
func (cs *CacheStore) Get(key string) *store.CacheEntry {
	cs.mu.RLock()
	entry, ok := cs.data[key]
	cs.mu.RUnlock()
	if !ok {
		entry = cs.load(key)
		if entry == nil {
			return nil
		}
	}
	if entry.TTL > 0 && entry.ExpiresAt != "" {
		expires, err := time.Parse(time.RFC3339, entry.ExpiresAt)
		if err == nil && time.Now().UTC().After(expires) {
			cs.mu.Lock()
			delete(cs.data, key)
			cs.mu.Unlock()
			cs.db.Exec(`DELETE FROM _cache WHERE key = ?`, key)
			return nil
		}
	}
	return entry
}

func (cs *CacheStore) load(key string) *store.CacheEntry {
	var e store.CacheEntry
	err := cs.db.QueryRow(`SELECT key, value, ttl, expires_at, created_at, updated_at FROM _cache WHERE key = ?`, key).
		Scan(&e.Key, &e.Value, &e.TTL, &e.ExpiresAt, &e.CreatedAt, &e.UpdatedAt)
	if err != nil {
		return nil
	}
	if e.TTL > 0 && e.ExpiresAt != "" {
		expires, err := time.Parse(time.RFC3339, e.ExpiresAt)
		if err == nil && time.Now().UTC().After(expires) {
			cs.db.Exec(`DELETE FROM _cache WHERE key = ?`, key)
			return nil
		}
	}
	cs.mu.Lock()
	cs.data[key] = &e
	cs.mu.Unlock()
	return &e
}

// Delete removes a key, reporting whether it was present.
func (cs *CacheStore) Delete(key string) bool {
	cs.mu.Lock()
	_, ok := cs.data[key]
	delete(cs.data, key)
	cs.mu.Unlock()
	cs.db.Exec(`DELETE FROM _cache WHERE key = ?`, key)
	return ok
}

// List returns all non-expired entries.
func (cs *CacheStore) List() []store.CacheEntry {
	now := time.Now().UTC()
	cs.mu.RLock()
	entries := make([]store.CacheEntry, 0, len(cs.data))
	for _, e := range cs.data {
		if e.TTL > 0 && e.ExpiresAt != "" {
			expires, err := time.Parse(time.RFC3339, e.ExpiresAt)
			if err == nil && now.After(expires) {
				continue
			}
		}
		entries = append(entries, *e)
	}
	cs.mu.RUnlock()
	return entries
}

// Flush removes all entries.
func (cs *CacheStore) Flush() {
	cs.mu.Lock()
	cs.data = make(map[string]*store.CacheEntry)
	cs.mu.Unlock()
	cs.db.Exec(`DELETE FROM _cache`)
}

// Stats reports in-memory cache usage.
func (cs *CacheStore) Stats() map[string]any {
	cs.mu.RLock()
	total := len(cs.data)
	now := time.Now().UTC()
	expired := 0
	for _, e := range cs.data {
		if e.TTL > 0 && e.ExpiresAt != "" {
			expires, err := time.Parse(time.RFC3339, e.ExpiresAt)
			if err == nil && now.After(expires) {
				expired++
			}
		}
	}
	cs.mu.RUnlock()
	return map[string]any{
		"total_entries":   total,
		"expired_entries": expired,
		"memory":          true,
	}
}

func (cs *CacheStore) persist(entry *store.CacheEntry) {
	cs.db.Exec(`INSERT INTO _cache (key, value, ttl, expires_at, created_at, updated_at) VALUES (?, ?, ?, ?, ?, ?)
		ON CONFLICT(key) DO UPDATE SET value = excluded.value, ttl = excluded.ttl, expires_at = excluded.expires_at, updated_at = excluded.updated_at`,
		entry.Key, entry.Value, entry.TTL, entry.ExpiresAt, entry.CreatedAt, entry.UpdatedAt)
}
